server/svix-server/src/db/models/applicationmetadata.rs

```diff
@@ -0,0 +1,95 @@
+// SPDX-FileCopyrightText: © 2022 Svix Authors
+// SPDX-License-Identifier: MIT
+use crate::core::types::ApplicationId;
+
+use crate::core::types::metadata::Metadata;
+use crate::{ctx, error};
+use chrono::Utc;
+use sea_orm::entity::prelude::*;
+use sea_orm::sea_query::OnConflict;
+use sea_orm::ActiveValue::Set;
+use sea_orm::ConnectionTrait;
+
+#[derive(Clone, Debug, PartialEq, Eq, DeriveEntityModel)]
+#[sea_orm(table_name = "applicationmetadata")]
+pub struct Model {
+    #[sea_orm(primary_key, auto_increment = false)]
+    pub id: ApplicationId,
+    pub created_at: DateTimeWithTimeZone,
+    pub updated_at: DateTimeWithTimeZone,
+    pub data: Metadata,
+}
+
+impl Model {
+    pub fn metadata(self) -> Metadata {
+        self.data
+    }
+}
+
+#[derive(Copy, Clone, Debug, EnumIter, DeriveRelation)]
+pub enum Relation {
+    #[sea_orm(
+        belongs_to = "super::application::Entity",
+        from = "Column::Id",
+        to = "super::application::Column::Id",
+        on_update = "NoAction",
+        on_delete = "Restrict"
+    )]
+    Application,
+}
+
+impl Related<super::application::Entity> for Entity {
+    fn to() -> RelationDef {
+        Relation::Application.def()
+    }
+}
+
+impl ActiveModelBehavior for ActiveModel {
+    fn before_save(mut self, _insert: bool) -> Result<Self, DbErr> {
+        self.updated_at = Set(Utc::now().into());
+        Ok(self)
+    }
+}
+
+impl ActiveModel {
+    pub fn create(app_id: ApplicationId, metadata: impl Into<Option<Metadata>>) -> Self {
+        let id = Set(app_id);
+        let data = Set(metadata.into().unwrap_or_default());
+        let timestamp = Utc::now();
+        Self {
+            id,
+            data,
+            created_at: Set(timestamp.into()),
+            updated_at: Set(timestamp.into()),
+        }
+    }
+
+    /// Upserts the record if it's new or updated, AND data is nonempty. Otherwise the record is
+    /// ignored or destroyed as appropriate.
+    pub async fn upsert_or_delete(self, db: &impl ConnectionTrait) -> error::Result<Metadata> {
+        let data = self.data.clone().take().unwrap_or_default();
+
+        // can't delete the record if there is no ID
+        if data.is_empty() && !self.id.is_not_set() {
```

Comment:
OK, I now see what was bothering me above, and this place shows it.

We can't on one hand say `create` requires an id, and on the other, have Metadata objects lying around that don't have `id` set.

Remove the `is_not_set`, and just assume it's always set. How to assume that? Remove the default implementation, and instead always create with a specific `id` when you create a fake one.

Though honestly, not sure about the value of creating a fake one always, it's fine to return `None` and handle it outside.

Reply:
I originally tried passing around `Option<Metadata>` / `Option<applicationmetadata::Model>` everywhere. It's fine from a correctness standpoint, but it _really_ adds a lot of cruft/boilerplate and conditional logic in the endpoints that are a huge PITA to deal with - and it makes the control flow a nightmare because you have to constantly worry about None values.

Just having a default `{}`, and then _one_ function that intelligently upserts is much simpler.

In practice `!self.id.is_not_set()` is probably never true - so I'll go ahead and removed the check.

Reply:
That is how it is currently written. I didn't override `ActiveModel::new()` - `ActiveModel::create(...)` ensures the ID is set.

We just have an optimization in the upsert where we don't insert empty data records. That's all this is, there's no extra risk of a runtime error than what was there before.